### PR TITLE
Initialize Timely lazily in computed

### LINKED ISSUE #14531: [Epic] Move Timely configuration to RPC protocol instead of relying on command line options
At the moment, a compute replica needs to be configured using command line arguments to set the number of workers, processes, peers, and the own process index. This is problematic because we need to pass information using two different mechanisms. Especially revealing the process index is based on a [hack](https://github.com/MaterializeInc/materialize/blob/main/src/compute/ci/entrypoint.sh#L16-L19) in Kubernetes.

The alternative is to move the configuration of the Timely parameters to the protocol that the compute controller and computed talk to each other. 

## Goals

* Couple the configuration of the Timely instance with the compute command protocol or the RPC interface.
* Avoid error-prone hacks to determine the process index in Kubernetes.

## Acceptance criteria

* The compute controller issues the Timely configuration parameters directly at a `computed` process instead of relying Kubernetes to forward the parameters.
* Reconciliation covers the configuration parameters of the Timely instance.

## Related

* #14379 attempted to implement this but ties the state of the Timely instance to the gRPC connection, which is not intended.

### PR DESCRIPTION
**Pull request by antiguru**
Initialize the Timely networking fabric using a dedicated `CreateTimely` command.

The command line arguments `--workers` and `--process` are removed from computed. 

### Motivation

This switches the initialization order of clusters around. Before, we would first initialize the Timely communication layer, and then start the GRPC server. With this change, the client provided to the GRPC server gains the knowledge on how to start and stop Timely clusters. This would be a prerequisite to the controller providing incarnation-specific UUIDs, which would allow the Timely fabric to detect split-brain situations.

The PR also removes access to secrets as a byproduct, they are not used anymore since removing general sinks from compute.

The PR removes the [hack](https://github.com/MaterializeInc/materialize/pull/14379/files#diff-8c02cfe28d853e3eeab6fea9b878eb410b865c017d1e013e0a218bb6f200102cL15) that used the hostname to determine the timely process id from the hostname.

Fixes #14531

## Tips to the reviewer

We can probably remove the `ClusterClientConfig` type, depending on how odd it feels.

The commands are documented [here](https://github.com/MaterializeInc/materialize/pull/14379/files#diff-8a737964bbd9cfe7847c5b19a1323b35ac048183d529f75d5e6cffc636acd200R44)

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered.
- [x] This PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way) and therefore is tagged with a `T-protobuf` label.
- [x] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):

  - None.  Developers that use `--unsafe-mode` must use the new remote replica syntax.

### CODE REVIEW

File: src/compute/src/server.rs
Comment (brennanvincent) [inline src/compute/src/server.rs:230]: It is not necessary to re-bind `cmd` in each branch

```suggestion
            ComputeCommand::CreateInstance(_) => {
                self.build()?;
                self.inner.as_mut().expect("intialized").send(cmd).await
            }
            ComputeCommand::DropInstance => {
                self.inner.as_mut().expect("intialized").send(cmd).await?;
                self.inner = None;
                let _ = self.worker_guards.take().map(|guards| guards.join());
                Ok(())
            }
            _ => self.inner.as_mut().expect("intialized").send(cmd).await,
```

File: src/compute/src/server.rs
Comment (brennanvincent) [inline src/compute/src/server.rs:89]: ?

**Review by brennanvincent — APPROVED** [review]

**Review by brennanvincent — CHANGES_REQUESTED** [review]
The crash in CI seems possibly legit/related (not sure).

```
platform-checks-computed_1-1       | thread 'timely:work-0' panicked at 'there is no reactor running, must be called from the context of a Tokio 1.x runtime', /var/lib/buildkite-agent/builds/buildkite-builders-d43b1b5-i-01dcb6c3e861741eb-1/materialize/tests/src/persist-client/src/internal/compact.rs:128:14
```

**Comment by brennanvincent** [comment]
NVM, I just found https://materializeinc.slack.com/archives/C01CFKM1QRF/p1661344441207669

File: src/compute/src/server.rs
Comment (antiguru) [reply src/compute/src/server.rs]: The rules of what match is a move and what's not are still surprising to me! Thanks, fixed.

File: src/compute/src/server.rs
Comment (antiguru) [reply src/compute/src/server.rs]: I removed the configuration type and inlined its fields.

Comment (brennanvincent) [reply src/compute/src/server.rs]: I mean you didn’t finish writing the doc comment on line 89

**Review by teskje — APPROVED** [review]
LGTM, just a couple nits. Although what's happening in src/compute/src/server.rs goes over my head, someone smarter than me needs to look at that!

**Comment by antiguru** [comment]
This PR has the defect that it ties the lifetime of the Timely instance to the gRPC connection, breaking all reconciliation on `computed`. We either need to reconcile the state of the Timely instance, or add another RPC call to initialize Timely outside the `ComputeCommand` protocol.

**Comment by lluki** [comment]
@teskje I added two commits (and fixed some in Moritz') that keep timely running on disconnects. It can probably be improved in various ways, but it passes the tests

File: src/compute-client/src/controller.rs
Comment (teskje) [inline src/compute-client/src/controller.rs]: "modified by"?

File: src/compute-client/src/command.rs
Comment (teskje) [inline src/compute-client/src/command.rs:48]: Can you explain in a comment (here or in some other place where the communication protocol is described) why we need the separate command, and why it cannot be bundled together with `CreateInstance`?

File: src/compute-client/src/command.proto
Comment (lluki) [reply src/compute-client/src/command.proto]: Protobuf encoding has now changed again, since `CreateTimely` and `CreateInstance` are now separate commands, `ProtoInstanceConfig` remains unchanged.

File: src/compute-client/src/command.rs
Comment (lluki) [reply src/compute-client/src/command.rs:74]: I improved the comments

File: misc/python/materialize/checks/actions.py
Comment (philip-stoev) [inline misc/python/materialize/checks/actions.py:44]: What is the difference between `REMOTE` and `COMPUTE` ?

Comment (lluki) [reply misc/python/materialize/checks/actions.py:44]: *  REMOTE says how environmentd connects to computed processes
* COMPUTE specifies to which address/port the timely connections are set up

If there is only one computed process, COMPUTE can be empty

File: test/replica-isolation/mzcompose.py
Comment (philip-stoev) [inline test/replica-isolation/mzcompose.py:78]: Can we omit all the `WORKERS 1` defaults from the mzcompose.py files and use a `1` default? The reason is that in the future I will need to introduce a default workers size setting (see https://github.com/MaterializeInc/materialize/pull/14272#issuecomment-1220356095) and all those `1` defaults interspersed throughout will be constantly overriding it.

Comment (antiguru) [reply test/replica-isolation/mzcompose.py:78]: Yes, I think this would make sense!

File: src/sql-parser/src/parser.rs
Comment (mjibson) [inline src/sql-parser/src/parser.rs:2880]: Can you add the new syntax to https://github.com/MaterializeInc/materialize/blob/main/src/sql-parser/tests/testdata/ddl?

File: src/adapter/src/catalog.rs
Comment (mjibson) [inline src/adapter/src/catalog.rs:4593]: This is persisted to the stash, so changing it requires either a migration or to wait for a wipe release to commit this PR. (We're aware we don't have automated tests for detecting this, it's on our priority list somewhere.)

I thiiink our current main is already a breaking release, so it might be safe to merge this already? You should check with, um, the #release channel first before merging though.

**Review by mjibson — APPROVED** [review]

File: src/sql-parser/src/parser.rs
Comment (lluki) [reply src/sql-parser/src/parser.rs:2901]: done

File: src/compute-client/src/command.rs
Comment (teskje) [inline src/compute-client/src/command.rs:54]:
```suggestion
///
```

File: src/compute-client/src/command.rs
Comment (teskje) [inline src/compute-client/src/command.rs:58]: Nit: `UpdateMaxResultSize` as well

File: src/compute-client/src/command.rs
Comment (teskje) [inline src/compute-client/src/command.rs:73]:
```suggestion
    /// ActiveReplication to all computed processes within a replica.
```

File: src/compute-client/src/controller.rs
Comment (teskje) [inline src/compute-client/src/controller.rs:698]: Is this new function used anywhere?

**Review by teskje — APPROVED** [review]
LGTM, apart from a couple nits and the `get_replica_ids` method that doesn't seem to be used anywhere (I guess it was removed in a parallel branch).

**Pull request merged**
